### ISSUE TITLE
Add persistent system-wide sideload-repos directory

### ISSUE THREAD

**Issue by mwleeds** [flatpak/flatpak#3688]
@jprvita suggested we consider adding a directory for system-wide configuration of `sideload-repos` (which is currently per-installation as well as in `/run`), which would allow the OS vendor to configure sideload repos without resorting to adding a link in each flatpak installation. Currently in Endless we have `/var/lib/flatpak/repo` be a link to `/ostree/repo`, the repo used for OS updates, but if we ever separate the two this would allow us to still sideload from `/ostree/repo` and avoid downloading any data there twice (though making the ostree repo a parent repo might be a better option to avoid on-disk duplication as well).

One potential pitfall to consider is what the behavior is if the same directory is linked to in more than place, so both in `/run/flatpak/sideload-repos/` and the persistent system-wide location.

It's also worth considering whether this should be implemented in libostree instead of flatpak.

**Comment by mwleeds** [comment]
> One potential pitfall to consider is what the behavior is if the same directory is linked to in more than place, so both in `/run/flatpak/sideload-repos/` and the persistent system-wide location.
> 

I think this is not a problem; libostree iterates over the list of localcache repos and uses the first one that has the object it needs and ignores the rest.
> It's also worth considering whether this should be implemented in libostree instead of flatpak.

On second thought it makes sense for the sideload-repos implementation to be in flatpak, because we want to decide what source to pull from in the resolve phase. For offline OS updates if we want to use a similar implementation we can still do that in ostree and eos-updater.

I'll remove the 1.8 milestone from this; I don't think it's urgent.

**Comment by ramcq** [comment]
> we want to decide what source to pull from in the resolve phase

I thought the resolve repos step was skipped as Flatpak is not using the full P2P API in ostree any more, so the sideload repos are passed just to ostree in "the" pull operation as local look-aside caches?

**Comment by mwleeds** [comment]
> > we want to decide what source to pull from in the resolve phase
> 
> I thought the resolve repos step was skipped as Flatpak is not using the full P2P API in ostree any more, so the sideload repos are passed just to ostree in "the" pull operation as local look-aside caches?

For each `FlatpakTransactionOperation` there's a `resolved_sideload_path` which we then pull from if set and don't use `localcache-repos` in that case. Perhaps @alexlarsson can speak to why we need to know where we're pulling from in the resolve phase? It looks like we at least use that information of whether we're using a sideload [to avoid downgrades](https://github.com/flatpak/flatpak/blob/2d2a05621a24a4c923087abe1b108cc6382aa1a4/common/flatpak-transaction.c#L2774).

As far as I can tell the `localcache-repos` pull option is only actually used for an install operation if the legacy `FlatpakInstallation` API was used instead of `FlatpakTransaction`. But doesn't that mean we can't use the USB repo to get some objects even if it provides something older than what's available on the Internet? I was under the impression that case was solved.

**Comment by mwleeds** [comment]
Previous discussion on this topic was here: https://github.com/flatpak/flatpak/pull/3476#issuecomment-601033201

I guess once we check the permissions metadata on a commit we want to pull that exact commit, but couldn't that be solved by specifying the commit to pull and still using `localcache-repos`?

**Comment by alexlarsson** [comment]
"Resolving" a ref is the action of finding the commit id that we are updating to (based on the ref name). If we have multiple repo sources (like upstream and one sideload, or two sideload) we need to look at all of them and find the one that has the most recent commit for that ref name. If this happens to be in a sideload repo, then that is remembered in transaction->sideload_path. 

On *top* of this, we always look in all the sideload repos for already existing objects, because they could be there even if the ref in that repo is older, or in fact if the ref doesn't even exist in that repo.

**Comment by alexlarsson** [comment]
> As far as I can tell the `localcache-repos` pull option is only actually used for an install operation if the legacy `FlatpakInstallation` API was used instead of `FlatpakTransaction`. But doesn't that mean we can't use the USB repo to get some objects even if it provides something older than what's available on the Internet? I was under the impression that case was solved.

Why do you think so? Its meant to always be set when we are pulling from "upstream". Its not set when we resolved the ref to a commit that is in a sideload repo, because then we have all data locally anyway, and there is no reason to look in some other local directory for files. (In theory one disk could be faster than another, but we do no such speed comparisons anywhere so it would be uselese.)

**Comment by mwleeds** [comment]
> > As far as I can tell the `localcache-repos` pull option is only actually used for an install operation if the legacy `FlatpakInstallation` API was used instead of `FlatpakTransaction`. But doesn't that mean we can't use the USB repo to get some objects even if it provides something older than what's available on the Internet? I was under the impression that case was solved.
> 
> Why do you think so? Its meant to always be set when we are pulling from "upstream". Its not set when we resolved the ref to a commit that is in a sideload repo, because then we have all data locally anyway, and there is no reason to look in some other local directory for files. (In theory one disk could be faster than another, but we do no such speed comparisons anywhere so it would be uselese.)

You're right, I was confusing resolving to a sideload path with resolving at all. Looks like we do use sideloads as caches when pulling from the Internet based on code inspection (haven't tested that experimentally).

**Comment by mwleeds** [comment]
> "Resolving" a ref is the action of finding the commit id that we are updating to (based on the ref name). If we have multiple repo sources (like upstream and one sideload, or two sideload) we need to look at all of them and find the one that has the most recent commit for that ref name. If this happens to be in a sideload repo, then that is remembered in transaction->sideload_path.
> 
The sideload path is remembered per-operation not per-transaction, which is good since different operations can resolve to different places.
> On _top_ of this, we always look in all the sideload repos for already existing objects, because they could be there even if the ref in that repo is older, or in fact if the ref doesn't even exist in that repo.

Not quite always. As discussed above we don't check other sideload repos in the case that a sideload provides the latest commit available for an operation.
